### PR TITLE
Nvidia mxnet

### PR DESCRIPTION
**Pull request by startakovsky**
@sc932, @pfhayes has approved this... please add your comments.

### CODE REVIEW

File: dnn-tuning-nvidia-mxnet/cnn_text/evaluate.py [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+    # For each training epoch
+    for iteration in range(epoch):
+        tic = time.time()
+        num_correct = 0
+        num_total = 0
+
+        # Over each batch of training data
+        for begin in range(0, x_train.shape[0], batch_size):
+            batchX = x_train[begin:begin+batch_size]
+            batchY = y_train[begin:begin+batch_size]
```
Comment (mikemccourt) [inline dnn-tuning-nvidia-mxnet/cnn_text/evaluate.py]: Technically, some Python people like spaces around their + signs.

File: dnn-tuning-nvidia-mxnet/cnn_text/evaluate.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+            norm = 0
+            for idx, weight, grad, name in cnn_model.param_blocks:
+                grad /= batch_size
+                l2_norm = mx.nd.norm(grad).asscalar()
+                norm += l2_norm * l2_norm
+
+            norm = np.sqrt(norm)
+            for idx, weight, grad, name in cnn_model.param_blocks:
+                if norm > max_grad_norm:
+                    grad *= (max_grad_norm / norm)
```
Comment (mikemccourt) [inline dnn-tuning-nvidia-mxnet/cnn_text/evaluate.py]: Parentheses are unnecessary here

File: dnn-tuning-nvidia-mxnet/mxnet-setup/install_mxnet_gpu_ubuntu.sh [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+sudo apt-get install -y git
+
+# Clone mxnet repository. In terminal, run the commands WITHOUT "sudo"
+git clone https://github.com/dmlc/mxnet.git ~/mxnet --recursive
+# If building with GPU, add configurations to config.mk file:
+cd ~/mxnet
+cp make/config.mk .
+echo "USE_CUDA=1" >>config.mk
+echo "USE_CUDA_PATH=/usr/local/cuda" >>config.mk
+echo "USE_CUDNN=1" >>config.mk
```
Comment (mikemccourt) [inline dnn-tuning-nvidia-mxnet/mxnet-setup/install_mxnet_gpu_ubuntu.sh]: inconsistent spacing

**Comment by alexandraj777** [comment]
@startakovsky Do you want to get a third party to test your setup script?

File: dnn-tuning-nvidia-mxnet/cnn_text/dataset.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
+"""
+Loads and preprocessed data for the MR dataset.
+Returns input vectors, labels, vocabulary, and inverse vocabulary.
+"""
+from collections import namedtuple
+
+import numpy as np
+
+from preprocess import (load_data_and_labels, pad_sentences,
```
Comment (alexandraj777) [inline dnn-tuning-nvidia-mxnet/cnn_text/dataset.py]: I thought Kevin fixed the indents?

**Comment by pfhayes** [comment]
I'm gonna test run through this now

File: dnn-tuning-nvidia-mxnet/cnn_text/graph.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
+import sys
+
+import mxnet as mx
```
Comment (pfhayes) [inline dnn-tuning-nvidia-mxnet/cnn_text/graph.py]: this import can take a long time... maybe we want a `print "Initializing..."` before all the imports, so users aren't frustrated by nothing happening

File: dnn-tuning-nvidia-mxnet/run_experiments.py [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+        param_filepath='cnn_text/hyperparams.json'
+
+    with open(param_filepath) as f:
+        hyperparams = f.read()
+        hyperparams = json.loads(hyperparams)
+
+    experiment = conn.experiments().create(
+                         name=exp_name,
+                         parameters=hyperparams,
+                         observation_budget=40*len(hyperparams))
```
Comment (pfhayes) [inline dnn-tuning-nvidia-mxnet/run_experiments.py]: why so high?

Comment (pfhayes) [inline dnn-tuning-nvidia-mxnet/run_experiments.py]: why so high?

**Comment by pfhayes** [comment]
Isn't running this going to take... days? anything we can do to at least inform users of that?

File: dnn-tuning-nvidia-mxnet/run_experiments.py (hunk 6 above)
Comment (pfhayes) [inline dnn-tuning-nvidia-mxnet/run_experiments.py]: can we add a `metric={"name":"Accuracy"}`?

File: dnn-tuning-nvidia-mxnet/cnn_text/dataset.py [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+    # Load and preprocess data
+    sentences, labels = load_data_and_labels()
+    sentences_padded = pad_sentences(sentences)
+    vocabulary, vocabulary_inv = build_vocab(sentences_padded)
+    x, y = build_input_data(sentences_padded, labels, vocabulary)
+
+    vocab_size = len(vocabulary)
+
+    # randomly shuffle data
+    np.random.seed(10)
```
Comment (alexandraj777) [inline dnn-tuning-nvidia-mxnet/cnn_text/dataset.py]: can you do this through a sklearn library? http://scikit-learn.org/stable/modules/cross_validation.html#random-permutations-cross-validation-a-k-a-shuffle-split Will this work?
